src/py/flwr/server/superlink/fleet/grpc_rere/server_interceptor.py

```diff
@@ -0,0 +1,176 @@
+# ==============================================================================
+"""Flower server interceptor."""
+
+
+import base64
+import threading
+from logging import INFO
+from typing import Any, Callable, Sequence, Set, Tuple, Union
+
+import grpc
+from cryptography.hazmat.primitives.asymmetric import ec
+
+from flwr.common.logger import log
+from flwr.common.secure_aggregation.crypto.symmetric_encryption import (
+    bytes_to_public_key,
+    generate_shared_key,
+    public_key_to_bytes,
+    verify_hmac,
+)
+from flwr.proto.fleet_pb2 import (  # pylint: disable=E0611
+    CreateNodeRequest,
+    CreateNodeResponse,
+    DeleteNodeRequest,
+    DeleteNodeResponse,
+    GetRunRequest,
+    GetRunResponse,
+    PullTaskInsRequest,
+    PullTaskInsResponse,
+    PushTaskResRequest,
+    PushTaskResResponse,
+)
+from flwr.server.superlink.state import StateFactory
+
+_PUBLIC_KEY_HEADER = "public-key"
+_AUTH_TOKEN_HEADER = "auth-token"
+
+Request = Union[
+    CreateNodeRequest,
+    DeleteNodeRequest,
+    PullTaskInsRequest,
+    PushTaskResRequest,
+    GetRunRequest,
+]
+
+Response = Union[
+    CreateNodeResponse,
+    DeleteNodeResponse,
+    PullTaskInsResponse,
+    PushTaskResResponse,
+    GetRunResponse,
+]
+
+
+def _get_value_from_tuples(
+    key_string: str, tuples: Sequence[Tuple[str, Union[str, bytes]]]
+) -> bytes:
+    value = next((value for key, value in tuples if key == key_string), "")
+    if isinstance(value, str):
+        return value.encode()
+
+    return value
+
+
+class AuthenticateServerInterceptor(grpc.ServerInterceptor):  # type: ignore
+    """Server interceptor for client authentication."""
+
+    def __init__(
+        self,
+        state_factory: StateFactory,
+        client_public_keys: Set[bytes],
+        private_key: ec.EllipticCurvePrivateKey,
+        public_key: ec.EllipticCurvePublicKey,
+    ):
+        self._lock = threading.Lock()
+        self.server_private_key = private_key
+        self.server_public_key = public_key
+        self.state = state_factory.state()
+        self.state.store_client_public_keys(client_public_keys)
+        self.encoded_server_public_key = base64.urlsafe_b64encode(
+            public_key_to_bytes(self.server_public_key)
+        )
+        log(
+            INFO,
+            "Client authentication enabled with %d known public keys",
+            len(client_public_keys),
+        )
+
+    def intercept_service(
+        self,
+        continuation: Callable[[Any], Any],
+        handler_call_details: grpc.HandlerCallDetails,
+    ) -> grpc.RpcMethodHandler:
+        """Flower server interceptor authentication logic.
+
+        Intercept unary call from client and do authentication process by validating
+        metadata sent from client. Continue RPC call if client is authenticated, else,
+        terminate RPC call by setting context to abort.
+        """
+        message_handler: grpc.RpcMethodHandler = continuation(handler_call_details)
```

Comment:
nit: should this be `method_handler`?

Reply:
@danieljanes the message_handler identifies that it's the default message_handler provided in flwr.server.superlink.fleet.message_handler to avoid confusion. Should I add this in a comment or do you prefer method_handler?

Reply:
@danielnugraha yes, let's add that as a comment for future readers